otherlibs/systhreads/st_stubs.c

```diff
@@ -470,7 +488,8 @@ CAMLprim value caml_thread_initialize(value unit)
   st_tls_newkey(&caml_thread_key);
 
   /* First initialise the systhread chain on this domain */
-  caml_thread_domain_initialize_hook();
+  if (!caml_thread_domain_initialize_hook())
+    caml_failwith("caml_thread_initialize: Thread initialization failed.");
```

Comment:
It would be really nice to distinguish between an OOM (`Out_of_memory`) and a thread allocation fail (`Failure`) here, if possible.

Reply:
Wish granted. Note that properly dealing with those exceptions for domains ≥ 1 will be more work, as mentioned.